### PR TITLE
Minor refactor ir with composite

### PR DESCRIPTION
**Pull request by shendiaomo**

### CODE REVIEW

File: pkg/sql/codegen/couler/codegen_test.go [hunk 1]
```diff
@@ -44,38 +44,8 @@ func mockSQLProgramIR() ir.SQLProgram {
...
-				"model.n_classes":    3},
-			Features: map[string][]ir.FeatureColumn{
-				"feature_columns": {
-					&ir.NumericColumn{&ir.FieldMeta{"sepal_length", ir.Float, "", []int{1}, false, nil, 0}},
-					&ir.NumericColumn{&ir.FieldMeta{"sepal_width", ir.Float, "", []int{1}, false, nil, 0}},
-					&ir.NumericColumn{&ir.FieldMeta{"petal_length", ir.Float, "", []int{1}, false, nil, 0}},
-					&ir.NumericColumn{&ir.FieldMeta{"petal_width", ir.Float, "", []int{1}, false, nil, 0}}}},
-			Label: &ir.NumericColumn{&ir.FieldMeta{"class", ir.Int, "", []int{1}, false, nil, 0}}},
-	}
+	trainStmt := ir.MockTrainStmt(fmt.Sprintf("mysql://%s", cfg.FormatDSN()), false)
```
Comment (typhoonzero) [inline pkg/sql/codegen/couler/codegen_test.go:47]: Cool!

File: pkg/sql/ir/ir.go [hunk 2]
```diff
@@ -76,17 +76,34 @@ type Executor interface {
...
+
+// IsExtended returns whether a SQLStatement is an extended SQL statement
+func (sql *ExtendedSQL) IsExtended() bool { return true }
+
+// GetOriginalSQL returns the original SQL statement used to get current IR result
+func (sql *ExtendedSQL) GetOriginalSQL() string { return sql.OriginalSQL }
+
+// TrainStmt is the intermediate representation for code generation of a training job.
+type TrainStmt struct {
+	ExtendedSQL
```
Comment (typhoonzero) [inline pkg/sql/ir/ir.go]: `TrainStmt. ExtendedSQL` seems confusing. Can direct implement `GetOriginalSQL` and `IsExtended` for each `TrainStmt`, `PredictStmt` and `AnalyzeStmt`

Comment (shendiaomo) [reply pkg/sql/ir/ir.go]: > `TrainStmt. ExtendedSQL` seems confusing. 

Structure embedding in Go has similar semantics as private inheritance in C++: they both disguise composition as inheritance. So we never need to write `TrainStmt.ExtendedSQL` or `TrainStmt::ExtendedSQL` explicitly. Struct embedding is good for extracting common fields/methods of several similar structs into a single struct.


> Can direct implement `GetOriginalSQL` and `IsExtended` for each `TrainStmt`, `PredictStmt` and `AnalyzeStmt`

I don't think code duplication is a good idea, maybe it's the naming of `ExtendedSQL` that causes the discomfort, how about renaming it as `ExtendedSQLBase`?  i.e. We have one `StandardSQL` and three `ExtendedSQL`s, the latter share implementation from `ExtendedSQLBase`.

**Review by wangkuiyi — DISMISSED** [review]
I have not looked deep into this change. I believe that changes that shorten code are good changes.

File: pkg/sql/ir/ir.go [hunk 3]
```diff
@@ -76,17 +76,34 @@ type Executor interface {
...
 	SetOriginalSQL(string)
 	Execute(Executor) error
+	IsExtended() bool
+	GetOriginalSQL() string
 }
 
-// TrainStmt is the intermediate representation for code generation of a training job.
-type TrainStmt struct {
+// ExtendedSQL contains common fields of TrainStmt, PredictStmt and AnalyzeStmt
+type ExtendedSQL struct {
```
Comment (Yancey0623) [inline pkg/sql/ir/ir.go]: The `ExtendedSQL` represent IR, do you want to rename `ExtendedSQL ` to `ExtendedStatment`

**Review by Yancey0623 — DISMISSED** [review]
LGTM with tiny comment.

**Review by typhoonzero — CHANGES_REQUESTED** [review]
Need more discussion about adding `ExtendedSQL` struct.

**Review by tonyyang-svail — APPROVED** [review]
LGTM.

**Comment by shendiaomo** [comment]
> Need more discussion about adding `ExtendedSQL` struct.

Revised as the results of the discussion.

**Review by typhoonzero — APPROVED** [review]

**Pull request merged**